packages/smart-contracts/src/contracts/SwapToPay.sol

```diff
@@ -0,0 +1,259 @@
+
+    /**
+     * @dev Emitted when the allowance of a `spender` for an `owner` is set by
+     * a call to {approve}. `value` is the new allowance.
+     */
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+}
+
+interface IFeePaymentNetwork {
+  // Event to declare a transfer with a reference
+  event TransferWithReferenceAndFee(
+    address tokenAddress,
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  );
+
+  function transferFromWithReferenceAndFee(
+    address _tokenAddress,
+    address _to,
+    uint256 _amount,
+    bytes calldata _paymentReference,
+    uint256 _feeAmount,
+    address _feeAddress
+    ) external;
+}
+
+/**
+ * -- ADMIN --
+ * 
+ * 1/ Deploy:
+ *    _uniswapRouterAddress: 0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D
+ *    _paymentProxyAddress: 0xda46309973bffddd5a10ce12c44d2ee266f45a44
+ * 
+ * 1/ This contract should allow the CTBK to be spent by the payment proxy
+ * approvePaymentProxyToSpend(0x995d6a8c21f24be1dd04e105dd0d83758343e258)
+ * 
+ * 2/ This contract should allow the FAU to be spent by the Uniswap router
+ * approveRouterToSpend(0xfab46e002bbf0b4509813474841e0716e6730136)
+ * 
+ * -- PAYER --
+ * 
+ * 1/ User should approve FAU to be spent by this contract
+ * Go here https://rinkeby.etherscan.io/token/0xfab46e002bbf0b4509813474841e0716e6730136#writeContract / approve 100 FAU
+ *    fau.approve(0x35177580d2f12c94cb274ec762121f4f7dfd451e, 100000000000000000000)
+ * 
+ * - Now start the actual payment -
+ * 
+ * EXEC transferMkrFromDaiWithReference
+ * _to: 0xAa0c45D2877373ad1AB2aa5Eab15563301e9b7b3
+ * _amount: 10000000000000000000 (10 FAU)
+ * _amountInMax: 10300300000000000000 (10.3003 CTBK = 3% slippage of the total amount)
+ * _path: [0x995d6a8c21f24be1dd04e105dd0d83758343e258, 0xfab46e002bbf0b4509813474841e0716e6730136]
+ * _paymentReference: 0x202009030000
+ * _feeAmount: 10000000000000000 (0.01 FAU = 0.1%)
+ * _feeAddress: 0x61076Da38517be36d433E3fF8D6875B87880Ba56
+ * _deadline: 1600136572 (15th of Sep)
+
+/**
+ * @title SwapToPay
+ * @notice This contract swaps ERC20 tokens before paying a request thanks to a payment proxy
+  */
+contract SwapToPay {
+
+  IUniswapV2Router02 public uniswapRouter;
+  IFeePaymentNetwork public paymentProxy;
+  address public admin;
+  
+
+  constructor(address _uniswapRouterAddress, address _paymentProxyAddress) public {
+    uniswapRouter = IUniswapV2Router02(_uniswapRouterAddress);
+    paymentProxy = IFeePaymentNetwork(_paymentProxyAddress);
+    admin = msg.sender;
+  }
+
+  function approvePaymentProxyToSpend(address erc20Address) public {
+    IERC20 erc20 = IERC20(erc20Address);
+    uint256 max = 2**256 - 1;
+    erc20.approve(address(paymentProxy), max);
+  }
+  
+  function approveRouterToSpend(address erc20Address) public {
+    IERC20 erc20 = IERC20(erc20Address);
+    uint256 max = 2**256 - 1;
+    erc20.approve(address(uniswapRouter), max);
```

Comment:
we should `require` here too

Reply:
Since USDT does not comply with the ERC20 require signature, and this function is useless if it fails, I suggest to leave it like this for now. OK for you?

Reply:
OK